### PR TITLE
locale: improve vc -> x11 keyboard conversion

### PR DESCRIPTION
**Pull request by fbuihuu**

### CODE REVIEW

File: src/locale/localed-util.c [hunk 1]
```diff
@@ -699,6 +697,38 @@ int vconsole_convert_to_x11(const VCContext *vc, X11Context *ret) {
...
+         * the conversion to the corresponding x11 layout is easy. */
+        _cleanup_free_ char *xlayout = NULL, *converted = NULL;
+        char *xvariant = NULL;
+
+        xlayout = strdup(vc->keymap);
+        if (!xlayout)
+                return -ENOMEM;
+        xvariant = strchr(xlayout, '-');
+        if (xvariant)
+                *xvariant++ = '\0';
```
Comment (poettering) [inline src/locale/localed-util.c]: this is unnecessarily hard to read. at least add `()` to make sure what is applied first. or better: just do two lines. no need to by unnecessarily cryptic

File: src/locale/localed-util.c [hunk 2]
```diff
@@ -699,6 +697,38 @@ int vconsole_convert_to_x11(const VCContext *vc, X11Context *ret) {
...
+
+        xc = (X11Context) {
+                .layout  = xlayout,
+                .model   = (char *) "microsoftpro",
```
Comment (poettering) [inline src/locale/localed-util.c]: hmm? where does this come from? needs a comment

File: src/locale/localed-util.c [hunk 3]
```diff
@@ -699,6 +697,38 @@ int vconsole_convert_to_x11(const VCContext *vc, X11Context *ret) {
...
+
+        /* No custom mapping has been found, see if it's the keymap is a converted one as in such case deduce
+         * the conversion to the corresponding x11 layout is easy. */
+        _cleanup_free_ char *xlayout = NULL, *converted = NULL;
+        char *xvariant = NULL;
+
+        xlayout = strdup(vc->keymap);
+        if (!xlayout)
+                return -ENOMEM;
+        xvariant = strchr(xlayout, '-');
```
Comment (poettering) [inline src/locale/localed-util.c:707]: might be a candidate for splitpair()

Comment (poettering) [reply src/locale/localed-util.c:707]: should this be strchr() btw? or strrchr()?

**Comment by poettering** [comment]
looks superficially ok to me, CI doesn't like it so. Will leave this to @keszybz to properly review

File: src/locale/localed-util.c (hunk 3 above)
Comment (fbuihuu) [reply src/locale/localed-util.c:707]: Looking at `/usr/share/kbd/keymaps/xkb/us-dvorak-intl.map.gz` file name, if `dvorak-intl` is really the variant here then `strchr()` I would say.

File: src/locale/localed-util.c (hunk 2 above)
Comment (fbuihuu) [reply src/locale/localed-util.c]: From the commit message:

> Note: by default the automatically generated x11 keyboard configs use keyboard
> model "microsoftpro" which should be equivalent to "pc105" model but with the
> internet/media key mapping added.

However if someone could confirm it that would be nice.

Comment (poettering) [reply src/locale/localed-util.c]: maybe add a comment about that to the sources, not just the commit msg

Comment (fbuihuu) [reply src/locale/localed-util.c]: Done

**Comment by yuwata** [comment]
```
426/1492 systemd:locale / test-localed-util                                                      FAIL              0.93s   exit status 1
>>> LD_LIBRARY_PATH=/systemd-meson-build/src/shared PROJECT_BUILD_ROOT=/systemd-meson-build MALLOC_PERTURB_=40 SYSTEMD_LANGUAGE_FALLBACK_MAP=/build/src/locale/language-fallback-map PATH=/systemd-meson-build:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl /systemd-meson-build/test-localed-util
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€• âœ€  â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•
stderr:
/* test_find_language_fallback */
/* test_find_converted_keymap */
Skipping rest of test_find_converted_keymap: keymaps are not installed
/* test_find_legacy_keymap */
Found legacy keymap pl2 with score 11
Found legacy keymap pl2 with score 6
/* test_vconsole_convert_to_x11 */
/* test empty keymap */
/* test without variant, new mapping (es:) */
/* test with known variant, new mapping (es:dvorak) */
/* test with unknown variant, new mapping (es:) */
/* test with unknown variant, new mapping (es:) step 1*/
/* test with unknown variant, new mapping (es:) step 2 */
../build/src/locale/test-localed-util.c:101:9: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/string.h:157:33: note: nonnull attribute specified here
    #0 0x55c03ff46d6b in test_vconsole_convert_to_x11 /systemd-meson-build/../build/src/locale/test-localed-util.c:101:9
    #1 0x55c03ff48996 in run_test_table /systemd-meson-build/../build/src/shared/tests.h:106:33
    #2 0x55c03ff48996 in main /systemd-meson-build/../build/src/locale/test-localed-util.c:196:1
    #3 0x7f8774e3c78f  (/usr/lib/libc.so.6+0x2378f) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #4 0x7f8774e3c849 in __libc_start_main (/usr/lib/libc.so.6+0x23849) (BuildId: 4a4bec3d95a1804443e852958fe59ed461135ce9)
    #5 0x55c03ff43614 in _start (/systemd-meson-build/test-localed-util+0xe614) (BuildId: 81f938905e58b0d7cddc6f5ab368f5b263d1a682)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../build/src/locale/test-localed-util.c:101:9 in 
```

**Comment by fbuihuu** [comment]
Thanks @yuwata, I was looking at the wrong logs which didn't such details.

So my suspicious is that `kbd` package is not installed on the test system, `/usr/share/kbd/keymaps/xkb/es.map.gz` is missing and therefore localed can't automatically find `es` layout and the new test fails.

Let's see if that's the case, updating the test to cope with this situation.

File: src/locale/localed-util.c [hunk 4]
```diff
@@ -697,6 +695,42 @@ int vconsole_convert_to_x11(const VCContext *vc, X11Context *ret) {
...
+        xc = (X11Context) {
+                .layout  = xlayout,
+                .model   = (char*) "microsoftpro",
+                .variant = xvariant,
+                .options = (char*) "terminate:ctrl_alt_bksp",
+        };
+
+        /* This sanity check seems redundant with the verification of the X11 layout done on the next
+         * step. However xkbcommon is an optional dependency hence the verification might be a NOP.  */
+        r = find_converted_keymap(&xc, &converted);
```
Comment (yuwata) [inline src/locale/localed-util.c:724]: Is it not necessary to check that `converted` is equivalent to `vc->keymap` ?

File: test/units/testsuite-73.sh [hunk 5]
```diff
@@ -273,15 +273,15 @@ test_vc_keymap() {
...
             assert_in "XKBOPTIONS=terminate:ctrl_alt_bksp" "$vc"
         elif [[ "$i" =~ ^us-.* ]]; then
-            assert_in "X11 Layout: .unset." "$output"
-            assert_not_in "X11 Model:" "$output"
-            assert_not_in "X11 Variant:" "$output"
-            assert_not_in "X11 Options:" "$output"
+            assert_in "X11 Layout: us" "$output"
+            assert_in "X11 Model: microsoftpro" "$output"
+            assert_in "X11 Variant:" "$output"
+            assert_in "X11 Options:i terminate:ctrl_alt_bksp" "$output"
```
Comment (yuwata) [inline test/units/testsuite-73.sh]: Please drop "i".

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks mostly good.

File: src/locale/localed-util.c (hunk 4 above)
Comment (fbuihuu) [reply src/locale/localed-util.c:724]: It shouldn't be necessary unless `find_converted_keymap()` is buggy. We could add an `assert_se()` to check it but that would slightly complexify the code, not sure whether it's worth it...

Comment (yuwata) [reply src/locale/localed-util.c:724]: I mean, `assert(streq_ptr(converted, vc->keymap));`. I think that is not complicated.

Comment (fbuihuu) [reply src/locale/localed-util.c:724]: well it depends on whether `find_converted_keymap()` could have found a keymap which has the variant.

Something like this ?

```
diff --git a/src/locale/localed-util.c b/src/locale/localed-util.c
index 02fac9786b..a67f65ecb8 100644
--- a/src/locale/localed-util.c
+++ b/src/locale/localed-util.c
@@ -722,10 +722,14 @@ int vconsole_convert_to_x11(const VCContext *vc, X11Context *ret) {
         /* This sanity check seems redundant with the verification of the X11 layout done on the next
          * step. However xkbcommon is an optional dependency hence the verification might be a NOP.  */
         r = find_converted_keymap(&xc, &converted);
+        if (r > 0)
+                assert_se(streq(vc->keymap, converted));
         if (r == 0 && xc.variant) {
                 /* If we still haven't find a match, try with no variant, it's still better than nothing.  */
                 xc.variant = NULL;
                 r = find_converted_keymap(&xc, &converted);
+                if (r > 0)
+                        assert_se(startswith(vc->keymap, converted));
         }
         if (r < 0)
                 return r;

```

Comment (yuwata) [reply src/locale/localed-util.c:724]: Yeah. Something like that. But, after reconsidering that, `assert_se()` may be too strong. How about to log something and refuse the conversion?
```
if (!streq(vc->keymap, converted)) {
        log_debug(...);
        *ret = (X11Context) {};
        return 0;
}
```

Comment (fbuihuu) [reply src/locale/localed-util.c:724]: honestly I don't see the point of such sanity check. The way `find_converted_keymap()` is to find a converted keymap whose filename matches the passed xlayout hence it's already ensured by the function itself.

Also why would we add such check here and not at all other call sites to `find_converted_keymap()` as well ?

Comment (yuwata) [reply src/locale/localed-util.c:724]: OK. That's just a minor suggestion. If you do not think it makes sense, then please ignore that.

**Pull request merged**